### PR TITLE
fix: trend detection for normalised time series

### PR DESCRIPTION
**Pull request by ChrisMarsden833**
Potential fix for [77](https://github.com/RussellSB/pytrendy/issues/77). Needs review.

---

Copilot summary:

This pull request introduces a debug mode for the trend detection pipeline and refines the method for detecting trends in time series data. The debug mode provides additional plots and print statements for developers, while the trend detection logic is improved by using a data-driven threshold based on the interquartile range (IQR) of the signal. The most important changes are grouped below:

**Debug Mode Enhancements:**

* Added a `debug` parameter to the `detect_trends` and `process_signals` functions, allowing developers to enable detailed debugging output, including additional plots and print statements. [[1]](diffhunk://#diff-94ff3b325f87933e0f2c545e7d19fb4cb819c29507918e5dde4d0d5b3f4fa9f8L11-R11) [[2]](diffhunk://#diff-4a8112027339d57e51021960187b86a9db7caf0c70137334303a3486a0c2a8b3R6-R9) [[3]](diffhunk://#diff-94ff3b325f87933e0f2c545e7d19fb4cb819c29507918e5dde4d0d5b3f4fa9f8R44-R47) [[4]](diffhunk://#diff-4a8112027339d57e51021960187b86a9db7caf0c70137334303a3486a0c2a8b3R35-R38) [[5]](diffhunk://#diff-94ff3b325f87933e0f2c545e7d19fb4cb819c29507918e5dde4d0d5b3f4fa9f8L65-R69)
* Implemented conditional plotting in `process_signals` that visualizes intermediate signal processing steps when `debug=True`, aiding in development and troubleshooting.

**Trend Detection Improvements:**

* Changed the trend detection threshold (`THRESHOLD_SMOOTH`) from a fixed value to a fraction of the signal's IQR, making the detection logic more adaptive to the data's variability. [[1]](diffhunk://#diff-4a8112027339d57e51021960187b86a9db7caf0c70137334303a3486a0c2a8b3L46-R51) [[2]](diffhunk://#diff-4a8112027339d57e51021960187b86a9db7caf0c70137334303a3486a0c2a8b3L176-R227)
* Updated the logic for assigning trend flags to use the new, data-driven threshold, improving robustness across different datasets.

### CODE REVIEW

File: pytrendy/process_signals.py [hunk 1]
```diff
@@ -173,38 +178,52 @@ def process_signals(df: pd.DataFrame, value_col: str) -> pd.DataFrame:
...
+
+    if debug:
+        import matplotlib.pyplot as plt
+
+        #df['smoothed_deriv'].hist()
+        #plt.show()
+
+        ax = df[[value_col, 'snr']].plot(figsize=(20,3), secondary_y='snr')
+        ax.right_ax.axhline(y=THRESHOLD_NOISE, color='gray', linestyle='--', linewidth=2)
+        plt.title("Signol-Noise Ratio (SNR)")
```
Comment (RussellSB) [inline pytrendy/process_signals.py]: Nitpick comment: smoll typo

File: pytrendy/process_signals.py [hunk 2]
```diff
@@ -31,6 +32,10 @@ def process_signals(df: pd.DataFrame, value_col: str) -> pd.DataFrame:
             Input time series data with a datetime index and signal column.
         value_col (str): 
             Name of the column containing the signal to process.
+        debug (bool, optional):
+            If `True` will run in debug mode, outputting various additional plots and print statements. 
+            Only recommended for developers of pytrendy.
+            Defaults to `False`.
```
Comment (RussellSB) [inline pytrendy/process_signals.py]: Nitpick comment: can you remove the line breaks between sentences here? and same for the copied docstring in detect_trends.py. This is so the auto api plugin of the docs can nicely format the argument description in the API Reference on the site docs. It follows a so-called Google style docstring format.

**Review by RussellSB — CHANGES_REQUESTED** [review]
Looks great. Thanks so much for this. I left some nitpick comments which shouldnt take much work. What might take some time though is adding tests.

For this change we will need to add 2 tests.

### Test 1

Test for the main focus of this commit, your [0,1] signal. 

- Could make a new module to add tests to called something like `tests/tests_normalised_trends` 
- You can reference your test data in a local csv somewhere under tests. E.g. save the csv `tests/tests_normalised_trends/normalised_data.csv` and have a test file that calls it `tests/test_normalised_trends.py`
- You can follow the similar pattern of most tests, e.g. `test_core_cases.py: line 18` . The idea is to load the signal, detect trends, and then assess that the detected segments are as expected.

### Test 2

Test for covering the lines of debug statements. 

- A bit silly, but there's hard enforced rules to maintain the maximized code coverage to avoid missing any new edge cases.
- This test could work with any data really .... To make things simple could use same test data above and just make sure that once debug True is enabled that it runs plt show calls from within process_trends(). 
- You can refer to `test_plot_pytrendy_edgecases.py: line 114` on how to do this with monkey patching. By setting detect trends plot=False, debug=True you should expect plt show to be hit from the debug True lines.

**Comment by ChrisMarsden833** [comment]
Just realised there are some other nitpick changes to be resolved here, will resolve these too.

**Comment by ChrisMarsden833** [comment]
I think I've addressed everything, see what you think.

File: pytrendy/__init__.py [hunk 3]
```diff
@@ -101,4 +101,5 @@
 from .detect_trends import detect_trends
 from .io.data_loader import load_data
 from .io.plot_pytrendy import plot_pytrendy
+from .io.results_pytrendy import PyTrendyResults
```
Comment (RussellSB) [inline pytrendy/__init__.py]: Nitpick: is this change necessary?

File: pytrendy/post_processing/segments_refine/abrupt_shaving.py [hunk 4]
```diff
@@ -42,8 +42,7 @@ def shave_abrupt_trends(df: pd.DataFrame, value_col: str, segments: list[dict],
             init_segment = init_segments[i]
             is_not_prev_trend = 'trend_class' not in init_segment # edge case, in case not trend before
             is_not_reclassified = is_not_prev_trend or segment['trend_class'] == init_segment['trend_class']
-            if is_not_reclassified:
```
Comment (RussellSB) [inline pytrendy/post_processing/segments_refine/abrupt_shaving.py:45]: Nitpick: is this change necessary?

File: tests/test_debug.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
+"""
+Tests for debug functionality.
+
+These tests verify that the debug process works as expected.
```
Comment (RussellSB) [inline tests/test_debug.py:4]: I'm not sure these equivalency tests are the best fit here. The logic from the monkey patch isn't getting used.

In 'test_plot_pytrendy_edgecases.py: line 114' , the assert method checks that the plt.show() gets called once from plot pytrendy as expected. I think if you rework this to be one test on gradual with debug enabled, then assert that the show calls == 7 (I believe that's the number of plt.show() calls in debug). That should be enough.

That way it will be testing that the debug code is working, that it plots the number of debug figures as expected.

**Review by RussellSB — CHANGES_REQUESTED** [review]
Hey, thanks Chris. The TestUncommonValues looks great. I have some feedback re the debug test. Also left some nitpick comments as I spotted some git diffs unrelated to the PR.

File: pytrendy/__init__.py (hunk 3 above)
Comment (ChrisMarsden833) [reply pytrendy/__init__.py]: It's necessary for type hinting the test functions; for the same reason I figured exposing it was a good idea, so people can type hint their own functions.

Can remove if desired.

File: pytrendy/post_processing/segments_refine/abrupt_shaving.py (hunk 4 above)
Comment (ChrisMarsden833) [reply pytrendy/post_processing/segments_refine/abrupt_shaving.py:45]: No, but I was getting 99% code coverage because of this line, it was annoying me.

File: tests/test_debug.py (hunk 5 above)
Comment (ChrisMarsden833) [reply tests/test_debug.py:4]: I didn't realise this - I thought it was just to suppress plot.show(). I'll look into fixing this as you suggest.

File: pytrendy/__init__.py (hunk 3 above)
Comment (RussellSB) [reply pytrendy/__init__.py]: Fair, but for that they could just do `from pytrendy.io.results_pytrendy import PyTrendyResults` and use that. 

By design, I'd rather keep the accessible methods in the main `__init__.py` specifically for user-facing functions

File: pytrendy/post_processing/segments_refine/abrupt_shaving.py (hunk 4 above)
Comment (RussellSB) [reply pytrendy/post_processing/segments_refine/abrupt_shaving.py:45]: Weird. The previous version with the linebreak achieves 100% code coverage in other recent PRs.

Maybe it was showing 99% locally cause not all tests were ran? Or a diff metric from codecov was referred to?

**Comment by ChrisMarsden833** [comment]
I've made the suggested changes and removed the nitpicks.

I've left in a pair of tests that check that debug mode doesn't affect the actual results, which feels particuarly important to include.

**Review by RussellSB — APPROVED** [review]
Yep looks good! Thanks for adding the monkey patch test with the num calls asserts.
Imo the equivalency tests could be removed since they aren't doing much, but not too imp to continue blocking, maybe could be cleaned up in a later PR.
